lib/TPP/ConvertTppToLoops.cpp

```diff
@@ -22,49 +22,75 @@ using namespace mlir::tpp;
 
 namespace {
 
+static bool isScalarVal(Value val) { return !val.getType().isa<ShapedType>(); }
+
+static SmallVector<Value> getLocalIvs(Value in, Value out, ValueRange ivs,
+                                      Value zero) {
+  assert(in.getType().isa<ShapedType>());
+  assert(out.getType().isa<ShapedType>());
+  auto shapedIn = in.getType().dyn_cast<ShapedType>();
+  auto shapedOut = out.getType().dyn_cast<ShapedType>();
+  assert(shapedOut.getRank() >= shapedIn.getRank());
+
+  // Handle rank 1 input.
+  auto shapeIn = shapedIn.getShape();
+  if (shapedIn.getRank() == 1) {
+    if (shapeIn[0] == 1)
+      return {zero};
+    return {ivs[shapedOut.getRank() - 1]};
+  }
+
+  // Handle rank 2 input.
+  assert(shapedIn.getRank() == shapedOut.getRank());
+  SmallVector<Value> newIvs;
+  for (auto [idx, shapeOnDim] : llvm::enumerate(shapeIn)) {
+    if (shapeOnDim == 1)
+      newIvs.push_back(zero);
+    else
+      newIvs.push_back(ivs[idx]);
+  }
+  return newIvs;
+}
+
+template <typename OpTy>
+static void convertTppToLoops(RewriterBase &rewriter, Location loc, Value lhs,
+                              Value rhs, Value out) {
+  assert(out.getType().isa<ShapedType>());
+  auto shape = out.getType().cast<ShapedType>().getShape();
+  int64_t rank = shape.size();
+  SmallVector<Value> ubs;
+  for (int64_t shapeDim : shape)
+    ubs.push_back(rewriter.create<arith::ConstantIndexOp>(loc, shapeDim));
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
```

Comment:
This pattern of zeroes and ones is *very* common, candidate to common up.

Reply:
done.